File: DAY-3/PART-2/code_2.py
import re

file_path = 'E:\ADVENT OF CODE\DAY-3\PART-2\code_2.txt'
with open(file_path, 'r') as file:
    memory = file.read()

pattern = re.compile(r"(do\(\)|don't\(\)|mul\((\d+),(\d+)\))")
matches = pattern.findall(memory)

total_sum = 0
enabled = True

for match in matches:
    instruction = match[0]
    
    if instruction == "do()":
        enabled = True
    elif instruction == "don't()":
        enabled = False
    else:
        if enabled:
            x = int(match[1])
            y = int(match[2])
            total_sum += x * y

print("Answer:", total_sum)

File: DAY-3/PART-2/code_2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>

void process_mul_operations(const char *filename) {
    FILE *file = fopen(filename, "r");  // Open file in read mode
    if (file == NULL) {
        perror("Error opening file");
        return;
    }

    char line[1024];
    regex_t regex;
    regmatch_t matches[3];

    // Compile regular expression to match mul(a, b)
    if (regcomp(&regex, "mul\\((\\d+),(\\d+)\\)", REG_EXTENDED)) {
        perror("Could not compile regex");
        fclose(file);
        return;
    }

    printf("Results of mul(a, b):\n");

    // Read the file line by line
    while (fgets(line, sizeof(line), file)) {
        int start = 0;
        // Search for the pattern mul(a, b)
        while (regexec(&regex, line + start, 3, matches, 0) == 0) {
            // Extract numbers a and b
            char a_str[64], b_str[64];
            int a, b;
            
            // Extract matched substrings
            snprintf(a_str, matches[1].rm_eo - matches[1].rm_so + 1, "%s", line + start + matches[1].rm_so);
            snprintf(b_str, matches[2].rm_eo - matches[2].rm_so + 1, "%s", line + start + matches[2].rm_so);
            
            // Convert extracted strings to integers
            a = atoi(a_str);
            b = atoi(b_str);
            
            // Print the multiplication result
            printf("mul(%d, %d) = %d\n", a, b, a * b);
            
            // Move the starting position to the next character after the current match
            start += matches[2].rm_eo;
        }
    }

    regfree(&regex);  // Free compiled regex
    fclose(file);     // Close the file
}

int main() {
    const char *filename = "code_2.txt";  // Specify the path to the input file
    process_mul_operations(filename);   // Call the function to process the file
    return 0;
}
